Sources/Task/ExistentialTask.swift

```diff
@@ -10,29 +10,35 @@
 import Deferred
 import Result
 #endif
-import Dispatch
-
-/// A simple function type expressing the cancellation of some operation.
-public typealias Cancellation = () -> Void
+import Foundation
 
 /// A wrapper over any task.
 ///
 /// Forwards operations to an arbitrary underlying future having the same result
 /// type, optionally combined with some `cancellation`.
-public struct Task<SuccessValue> {
+public final class Task<SuccessValue>: NSObject, NSProgressReporting {
     public typealias Result = TaskResult<SuccessValue>
 
     private let future: Future<Result>
-    private let cancellation: Cancellation
+    public let progress: NSProgress
 
-    /// Creates a task given a `future` and an optional `cancellation`.
-    public init(_ future: Future<Result>, cancellation: Cancellation) {
+    /// Creates a task given a `future` and its `progress`.
+    public init(future: Future<Result>, progress: NSProgress) {
         self.future = future
-        self.cancellation = cancellation
+        self.progress = NSProgress.extendingRoot(for: progress)
```

Comment:
Maybe `extendingRoot` should move into the `map`, `flatMap`, or `recover`. This design is flawed for allowing custom implementers of `TaskType`.